File: problem-solving/A2SV/leetcode/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        dic = defaultdict(int)
        stack = []
        for i in range(len(s)-1,-1,-1):
            if s[i] == ")" or s[i] == "}" or s[i] == "]":
                dic[s[i]] += 1
                stack.append(s[i])
            elif s[i] == "(":
                if dic[")"] > 0 and stack and stack[-1] == ")":
                    dic[")"] -= 1
                    stack.pop()
                else:
                    return False
            elif s[i] == "{":
                if dic["}"] > 0 and stack and stack[-1] == "}":
                    dic["}"] -= 1
                    stack.pop()
                else:
                    return False
            elif s[i] == "[":
                if dic["]"] > 0 and stack and stack[-1] == "]":
                    dic["]"] -= 1
                    stack.pop()
                else:
                    return False
        return max(dic.values()) == 0 
            

File: 20-valid-parentheses/valid-parentheses.go
func isValid(s string) bool {
    stack := []string{}
    for _,x := range s{
        if string(x) == ")"{
            if len(stack) > 0 && stack[len(stack)-1] == "("{
                stack = stack[:len(stack)-1]
            } else {
                return false
            }
        } else if string(x) == "}" {
            if len(stack) > 0 && stack[len(stack)-1] == "{"{
                stack = stack[:len(stack)-1]
            } else {
                return false
            }
        } else if string(x) == "]" {
            if len(stack) > 0 && stack[len(stack)-1] == "["{
                stack = stack[:len(stack)-1]
            } else {
                return false
            }
        } else {
            stack = append(stack,string(x))
        }

    }
    return len(stack) == 0
}